hardware_interface/include/hardware_interface/resource_manager.hpp

```diff
@@ -0,0 +1,115 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef HARDWARE_INTERFACE__RESOURCE_MANAGER_HPP_
+#define HARDWARE_INTERFACE__RESOURCE_MANAGER_HPP_
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "hardware_interface/actuator_hardware_interface.hpp"
+#include "hardware_interface/actuator_hardware.hpp"
+#include "hardware_interface/robot_hardware_interface.hpp"
+#include "hardware_interface/sensor_hardware_interface.hpp"
+#include "hardware_interface/sensor_hardware.hpp"
+#include "hardware_interface/system_hardware_interface.hpp"
+#include "hardware_interface/system_hardware.hpp"
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "hardware_interface/visibility_control.h"
+#include "pluginlib/class_loader.hpp"
+
+using hardware_interface::return_type;
+
+namespace resource_manager
+{
+
+class ResourceManager
+{
+public:
+  HARDWARE_INTERFACE_PUBLIC
+  ResourceManager();
+
+  HARDWARE_INTERFACE_PUBLIC
+  virtual
+  ~ResourceManager() = default;
+
+  //  Non real-time safe functions
+  HARDWARE_INTERFACE_PUBLIC
+  return_type load_and_configure_resources_from_urdf(std::string urdf_string);
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type check_command_interfaces(
+    const std::string & joint_name, const std::vector<std::string> & interfaces) const;
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type claim_command_handle(
+    const std::string & joint_name, const std::vector<std::string> & interfaces,
+    std::shared_ptr<hardware_interface::components::Joint> & command_handle);
+
+  //  Real-time safe functions
+  HARDWARE_INTERFACE_PUBLIC
+  return_type start_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type stop_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type read_all_resources();
+
+  HARDWARE_INTERFACE_PUBLIC
+  return_type write_all_resources();
+
+private:
+  // TODO(all): make this unique?
+  std::vector<std::shared_ptr<hardware_interface::ActuatorHardware>> actuators_;
+  std::vector<std::shared_ptr<hardware_interface::SensorHardware>> sensors_;
+  std::vector<std::shared_ptr<hardware_interface::SystemHardware>> systems_;
+
+  std::map<std::string, std::shared_ptr<
+      hardware_interface::components::Joint>> joint_components_;
+  std::map<std::string, std::vector<std::shared_ptr<
+      hardware_interface::components::Joint>>> joint_components_for_hardware_;
+  std::map<std::string, std::shared_ptr<
+      hardware_interface::components::Sensor>> sensor_components_;
+  std::map<std::string, std::vector<std::shared_ptr<
+      hardware_interface::components::Sensor>>> sensor_components_for_hardware_;
+
+  /**
+   * Map of joints and their command interfaces
+   */
+  std::map<std::string, std::vector<std::string>> command_interfaces_;
+
+  std::map<std::string, std::vector<std::string>> claimed_command_interfaces_;
```

Comment:
why do we need a map?

Reply:
And if needed, consider using a unordered_map for speed.

Reply:
The maps are here so one can easily find the objects based on their name.

so `claimed_command_interfaces_` is storing joints and a list of claimed interface. So we can claim different interfaces by different controllers.

Reply:
The map type is changed. I would keep this for now until we clarified controller initialization in #183